Source/WebCore/html/closewatcher/CloseWatcher.cpp

```diff
@@ -0,0 +1,149 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CloseWatcher.h"
+
+#include "CloseWatcherManager.h"
+#include "Event.h"
+#include "EventNames.h"
+#include "KeyboardEvent.h"
+#include "LocalDOMWindow.h"
+#include "ScriptExecutionContext.h"
+#include <wtf/IsoMallocInlines.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(CloseWatcher);
+
+ExceptionOr<Ref<CloseWatcher>> CloseWatcher::create(ScriptExecutionContext& context, const CloseWatcher::Options& options)
+{
+    RefPtr document = dynamicDowncast<Document>(&context);
+    if (!document || !document->domWindow())
+        return Exception { ExceptionCode::InvalidStateError, "CloseWatchers cannot be created in a detached window."_s };
+    if (!document->isFullyActive())
+        return Exception { ExceptionCode::InvalidStateError, "Document is not fully active."_s };
+
+    Ref watcher = CloseWatcher::establish(*document);
+
+    if (options.signal) {
+        if (options.signal->aborted()) {
+            watcher->m_closed = true;
+            document->domWindow()->closeWatcherManager().remove(watcher.get());
+        } else {
+            watcher->m_signal = options.signal;
+            watcher->m_signalAlgorithm = options.signal->addAlgorithm([weakWatcher = WeakPtr { watcher.get() }](JSC::JSValue) mutable {
+                if (weakWatcher)
+                    weakWatcher->destroy();
+            });
+        }
+    }
+
+    return watcher;
+}
+
+Ref<CloseWatcher> CloseWatcher::establish(Document& document)
+{
+    ASSERT(document.isFullyActive());
+
+    Ref<CloseWatcher> watcher = adoptRef(*new CloseWatcher(document));
+    watcher->suspendIfNeeded();
+
+    auto& manager = document.domWindow()->closeWatcherManager();
+
+    manager.add(watcher);
+    return watcher;
+}
+
+CloseWatcher::CloseWatcher(Document& document)
+    : ActiveDOMObject(document)
+{ }
+
+void CloseWatcher::requestClose()
+{
+    requestToClose();
+}
+
+bool CloseWatcher::requestToClose()
```

Comment:
Because the web exposed method has a return type of void but the internal algorithm has a return type of bool. If I just try to have the one method which returns bool the build fails.

```
JSDOMConvertBase.h:148:42: error: no viable conversion from
      'bool' to 'std::monostate'
  148 |         return JSConverter<IDL>::convert(std::forward<U>(value));
  ```